createReviewerMapping: impossible override due to wrong array key

krisElphenor opened the pull request:
Please see issue https://github.com/Sylius/Sylius/issues/7471

| Q               | A
| --------------- | ---
| Bug fix?        | yes |
| New feature?    | no |
| BC breaks?      | no |
| Related tickets |  mentioned in #7471 |
| License         | MIT |

lchrusciel commented on src/Sylius/Bundle/ReviewBundle/Doctrine/ORM/Subscriber/LoadMetadataSubscriber.php:
```diff
@@ -106,11 +106,13 @@ private function createReviewerMapping($reviewerEntity, ClassMetadata $reviewerE
-                'referencedColumnName' => $reviewerEntityMetadata->fieldMappings['id']['columnName'],
-                'nullable' => false,
-                'onDelete' => 'CASCADE',
+            'joinColumns' => [
```
We used to put both arrays openings in the same line (check [LoadMetadataSubscriber in attribute bundle](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/AttributeBundle/Doctrine/ORM/Subscriber/LoadMetadataSubscriber.php#L85-L90)).

krisElphenor replied:
Ok, no worries. But, in that case, to be consistent, we should also update the `createSubjectMapping` method in that same file, as it has the same formatting (see https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/ReviewBundle/Doctrine/ORM/Subscriber/LoadMetadataSubscriber.php#L87).

If you wish, I can update this PR to include that changes for both method.

lchrusciel replied:
If you could. Anyway, you need to generate a migration file and update our specs (that's why test suites are failing)

lchrusciel commented:
And we still need a doctrine migration 
```bash
$ php app/console doctrine:migration:migrate
```

krisElphenor commented:
I've added the migrations and updated the format and the specs. Sorry for the delay and the multiple commit, I'll try to make it more smooth next time :)

krisElphenor commented:
There seems to be an issue with the scenario, I'll check it...

krisElphenor commented:
The problem comes from the condition `And this product has one review` which is satisfied by the method `productHasAReview` defined here: [ProductReviewContext](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Behat/Context/Setup/ProductReviewContext.php#L70)

In this method, the call to `createProductReview` is allowed to set the author of the review as null, which is not allowed by the mapping. 

So, I guess the question is: do we allow anonymous reviews or not ?

pamil commented on src/Sylius/Bundle/ReviewBundle/Doctrine/ORM/Subscriber/LoadMetadataSubscriber.php:
```diff
@@ -106,12 +104,12 @@ private function createReviewerMapping($reviewerEntity, ClassMetadata $reviewerE
                 'name' => 'author_id',
                 'referencedColumnName' => $reviewerEntityMetadata->fieldMappings['id']['columnName'],
-                'nullable' => false,
+                'nullable' => true,
```
Why?

krisElphenor replied:
Because of my previous comment. 

> In this method, the call to createProductReview is allowed to set the author of the review as null, which is not allowed by the mapping.

So, I've updated the mapping to follow the code expectation, but if a review should always have an author, then this should, indeed, be set to non-nullable and the code need to be updated to reflect that as well.

pamil replied:
/cc @Zales0123 @GSadee

Zales0123 replied:
Yes, review definitely should have an author in all cases, so it should not be nullable.

michalmarcinkowski commented:
@krisElphenor we should not allow for `null` value as author and the step  `And this product has one review` should be renamed to `And this product has one review from "john@doe.com"` so we can set the author.

Zales0123 commented on app/migrations/Version20170215115804.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+class Version20170215115804 extends AbstractMigration
```
Could you please merge this two migrations in one?

krisElphenor commented:
I have done the changes needed. Unfortunately, it seems to fail on 'adding a product with association'. 
I'm quite new to behat, so, i'm sorry for all those little steps and if i'm missing something obvious, but I don't see the link between the associations and reviews. So what do you think I'm missing ?

lchrusciel commented on features/product/managing_products/deleting_product.feature:
```diff
@@ -8,7 +8,7 @@ Feature: Deleting a product
         And the store has a product "Toyota GT86 model"
         And this product has "1:43" variant priced at "$15.00"
-        And this product has one review
+        And this product has one review from "john@doe.com"
```
you have missed `customer`.

And this product has one review from **customer** "john@doe.com"

krisElphenor replied:
Done ! Thanks for the help @lchrusciel :)

michalmarcinkowski commented:
@krisElphenor could you try the fix proposed by @lchrusciel?

pjedrzejewski commented:
Thanks @krisElphenor, nice work!

krisElphenor commented:
No worries, but more importantly, thanks to all of you for the great work you do on Sylius.
